Feat(eos_cli_config_gen)!: Add support for multiple virtual router addresses for vlans

tgodaA opened the pull request:
- Modified design: In Network services, ip_virtual_router_address changed to ip_virtual_router_addresses to support multiple virtual router addresses. The new data type must be a list!
- Added molecule tests for eos_designs_unit_tests and eos_cli_config_gen
- Added design changes to release notes 3.x.x.md
- Modified documentation template in eos_cli_config_gen role to display all virtual-router IP addresses, used the join(", ") function over the array.

## Change Summary
eos_designs, eos_cli_config_gen

- Affected roles: eos_designs, eos_cli_config_gen. The vlan-interfaces.j2 is modified in both of the roles. 
- eos_designs template path: ansible-avd/ansible_collections/arista/avd/roles/eos_designs/templates/designs/l3ls-evpn/tenants/vlan-interfaces.j2. Existing data model is modified.
- eos_cli_config_gen template path: ansible-avd/ansible_collections/arista/avd/roles/eos_cli_config_gen/templates/eos/vlan-interfaces.j2. Existing snippet is modified to support multiple IP addresses.
- Changes made to following eos_designs templates: static-routes, inband_management to make sure the static route for VARP (with IPv4/Mask type) are added to the SVI. 

<!-- Enter short PR description -->

## Related Issue(s)

Fixes #1167 

## Component(s) name

`eos_designs`, `eos_cli_config_gen`

## Proposed changes
<!--- Describe your changes in detail -->
- tenants.{{ tenant }}.vrfs.{{ vrf }}.svis.{{ svi }}.ip_virtual_router_address **to** tenants.{{ tenant }}.vrfs.{{ vrf }}.svis.{{ svi }}.ip_virtual_router_addresses
- svi_profiles.{{ profile }}.ip_virtual_router_address **to** svi_profiles.{{ profile }}.ip_virtual_router_addresses
<!--- Describe data model implemented for new features -->

## How to test
<!--- Please describe in detail how you tested your changes. -->
group_vars/DC1_TENANTS_NETWORKS.yml:
```
tenants:
  # Tenant A Specific Information - VRFs / VLANs
  Tenant_A:
    vrfs:
      Tenant_A_OP_Zone:
        vrf_vni: 10
        vtep_diagnostic:
          loopback: 100
          loopback_ip_range: 10.254.1.0/24
        svis:
          12:
            name: Tenant_A_OP_Zone_Test
            tags: [opzone]
            enabled: true
            ip_virtual_router_addresses: [<ipv4/mask>, <ipv4>] 
```
<!--- Include details of your testing environment, and the tests you ran to -->

## Checklist

### User Checklist

<!-- Add your own checklist using MD syntax and by replacing N/A -->
- N/A

### Repository Checklist

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ x] My code has been rebased from devel before I start
- [ x] I have read the [**CONTRIBUTING**](https://avd.sh/en/latest/docs/contribution/overview.html) document.
- [ x] My change requires a change to the documentation and documentation have been updated accordingly.
- [ x] I have updated [molecule CI](https://github.com/aristanetworks/ansible-avd/tree/devel/ansible_collections/arista/avd/molecule) testing accordingly. (check the box if not applicable)

ClausHolbechArista commented on ansible_collections/arista/avd/roles/eos_designs/doc/l3ls-evpn/network-services.md:
```diff
@@ -45,7 +45,9 @@ svi_profiles:
     enabled: < true | false >
-    ip_virtual_router_address: < IPv4_address/Mask >
+    ip_virtual_router_addresses:
+      - < IPv4_address/Mask >
```
Note there is a big difference in EOS between `ip virtual-router address <ip / mask>` and `ip virtual-router address <ip>`. The template currently has a filter that removes any mask, so either we should only show "IPv4_address" here or we change the template to _not_ extract the IP alone, and then show both options `< IPv4_address | IPv4Address/Mask >` here.

tgodaA replied:
My bad, it was a wrong description. I will remove /Mask as it should not have any mask.

ClausHolbechArista commented on ansible_collections/arista/avd/roles/eos_designs/templates/designs/l3ls-evpn/tenants/vlan-interfaces.j2:
```diff
@@ -49,8 +49,8 @@ vlan_interfaces:
-{%             if svi_ip_virtual_router_address is arista.avd.defined %}
-    ip_virtual_router_address: {{ svi_ip_virtual_router_address | ansible.netcommon.ipaddr('address') }}
+{%             if svi_ip_virtual_router_addresses is arista.avd.defined %}
+    ip_virtual_router_addresses: {{ svi_ip_virtual_router_addresses | ansible.netcommon.ipaddr('address') }}
```
This filter removes the mask.

ClausHolbechArista replied:
I think we should remove this extra `ansible.netcommon.ipaddr('address')`filter so we can accept `< IPv4 Address>` _and_ `< IPv4 Address/Mask >`

ClausHolbechArista commented:
Remember to also patch the inband management templates with this new key name. The failing tests you have are caused by that.

ClausHolbechArista commented on ansible_collections/arista/avd/molecule/eos_cli_config_gen/intended/configs/vlan-interfaces.cfg:
```diff
@@ -53,11 +53,13 @@ interface Vlan83
 interface Vlan84
    description SVI Description
    ip address 10.10.84.1/24
-   ip virtual-router address 10.10.84.254
```
This line is being removed with your changes. We should never accept things disappearing from molecule unless we intended so. In this case you should remember to update the data model for vlan84 as well.

ClausHolbechArista commented on ansible_collections/arista/avd/molecule/eos_designs_unit_tests/inventory/group_vars/DC1_TENANTS_NETWORKS.yml:
```diff
@@ -70,7 +70,17 @@ tenants:
             name: Tenant_A_WEBZone_2
             tags: ['web']
-            profile: GENERIC_FULL
+            nodes:
```
Please don't remove stuff from existing Vlans. It will remove test coverage of other features. Add a new Vlan to test your combination unless it can be done without impact to the existing tests

ClausHolbechArista commented on ansible_collections/arista/avd/roles/eos_cli_config_gen/templates/documentation/vlan-interfaces.j2:
```diff
@@ -36,7 +36,7 @@
 | --------- | --- | ---------- | ------------------ | ------------------------- | ---- | ------ | ------- |
 {%     for vlan_interface in vlan_interfaces | arista.avd.natural_sort %}
-| {{ vlan_interface }} | {% if vlan_interfaces[vlan_interface].vrf is defined and vlan_interfaces[vlan_interface].vrf is not none %} {{ vlan_interfaces[vlan_interface].vrf }} {% else %} default {% endif %} | {% if vlan_interfaces[vlan_interface].ip_address is defined and vlan_interfaces[vlan_interface].ip_address is not none %} {{ vlan_interfaces[vlan_interface].ip_address }} {% else %} - {% endi ...
+| {{ vlan_interface }} | {% if vlan_interfaces[vlan_interface].vrf is defined and vlan_interfaces[vlan_interface].vrf is not none %} {{ vlan_interfaces[vlan_interface].vrf }} {% else %} default {% endif %} | {% if vlan_interfaces[vlan_interface].ip_address is defined and vlan_interfaces[vlan_interface].ip_address is not none %} {{ vlan_interfaces[vlan_interface].ip_address }} {% else %} - {% endi ...
```
We try to avoid inline jinja, since it is hard to read. Please take a look at some of the other documentation templates, where we calculate all values first and then print out the line.
I understand you are only changing an existing template, but we should try to improve the templates when we touch them. Let me know if you need a hand on this.

ClausHolbechArista commented on ansible_collections/arista/avd/roles/eos_cli_config_gen/templates/documentation/vlan-interfaces.j2:
````diff
@@ -86,4 +86,4 @@
 {% include 'eos/vlan-interfaces.j2' %}
 ```
-{% endif %}
+{% endif %}
````
Missing newline at the end of the file

ClausHolbechArista commented on ansible_collections/arista/avd/roles/eos_designs/doc/l3ls-evpn/network-services.md:
```diff
@@ -158,8 +161,10 @@ tenants:
             # note, also requires an IP address to be configured on the SVI where it is applied.
-            # Optional
-            ip_virtual_router_address: < IPv4_address/Mask >
+            # Optional, don't assign mask else that IP will be rejected
```
Are you sure it will be rejected? The mask will be removed by the filter, but not the IP will survive.
I think we should actually remove that filter in the templates, so we accept mask on this as well. There are valid use cases for that, and it will improve our data model coverage.

ClausHolbechArista reviewed (dismissed):
A few comments. Let us sync up on chat.

tgodaA commented:
I will resubmit the changes once again. My intention was to add a comment to not include subnets, https://docs.ansible.com/ansible/2.4/playbooks_filters_ipaddr.html#getting-information-from-host-prefix-values.

My bad I interpreted the message wrong.

tgodaA committed:
Feat(eos_cli_config_gen): Add support for multiple virtual router addresses for vlans

Modified designs and files:
- In Network services, ip_virtual_router_address changed to ip_virtual_router_addresses to support multiple virtual router addresses. The new data type must be a list!
- Added molecule tests for eos_designs_unit_tests and eos_cli_config_gen
- Added design changes to release notes 3.x.x.md
- Modified documentation template in eos_cli_config_gen role to display all virtual-router IP addresses, used the join(", ") function over the array.
- Made changes to inband_management
- Made changes to static-routes templates, to add static_routes for IP_Virtual_Router_Address/mask VARP.

Fixes #1167

tgodaA committed:
Made changes to inband_management template. Added ipaddr('address') filter to remove prefix from the IP address. Ran molecule scenarios with eos_designs in its name and updated all molecule scenarios to this commit.

tgodaA committed:
Updated README.md

Updated ip_virtual_router_address with ip_virtual_router_addresses in the README file.

carlbuchmann reviewed (approved):
LGTM!

titom73 reviewed (approved):
LGTM!

carlbuchmann reviewed (approved):
LGTM, all comments address by @tgodaA
